pkg/csi/service/vanilla/controller.go

```diff
@@ -1243,14 +1244,75 @@ func (c *controller) CreateSnapshot(ctx context.Context, req *csi.CreateSnapshot
 				"cannot snapshot migrated vSphere volume. :%q", volumeID)
 		}
 
-		// Query Capacity in MB for block volume snapshot
-		snapshotSizeInMB, err := utils.QueryCapacityForBlockVolumeSnapshotUtil(
+		// Query capacity in MB and datastore url for block volume snapshot
+		snapshotSizeInMB, datastoreUrl, err := utils.QueryCapacityAndDsUrlForVolumeUtil(
 			ctx, c.manager.VolumeManager, volumeID, common.BlockVolumeType)
 		if err != nil {
 			return nil, err
 		}
 		volumeType = prometheus.PrometheusBlockVolumeType
 
+		// Check if snapshots number of this volume reaches the granular limit on VSAN/VVOL
+		maxSnapshotsPerBlockVolume := c.manager.CnsConfig.Snapshot.GlobalMaxSnapshotsPerBlockVolume
+		log.Infof("The limit of the maximum number of snapshots per block volume is "+
+			"set to the global maximum (%v) by default.", maxSnapshotsPerBlockVolume)
+		if c.manager.CnsConfig.Snapshot.GranularMaxSnapshotsPerBlockVolumeInVSAN != 0 ||
+			c.manager.CnsConfig.Snapshot.GranularMaxSnapshotsPerBlockVolumeInVVOL != 0 {
+			// query the datastore type of the source volume
+			vCenter, err := cnsvsphere.GetVirtualCenterManager(ctx).GetVirtualCenter(ctx, c.manager.VcenterConfig.Host)
+			if err != nil {
+				return nil, logger.LogNewErrorCodef(log, codes.Internal,
+					"failed to get vCenter when checking if the number of snapshots on the source volume "+
+						"reaches the configured maximum. err: %v", err)
+			}
+
+			dsInfo, err := cnsvsphere.GetDatastoreInfoByURL(
+				ctx, vCenter, c.manager.CnsConfig.Global.ClusterID, datastoreUrl)
+			if err != nil {
+				return nil, logger.LogNewErrorCodef(log, codes.Internal,
+					"failed to get the datastore info corresponding to URL %q when checking if the number of "+
+						"snapshots on the source volume reaches the configured maximum. err: %v", datastoreUrl, err)
+			}
+
+			dsType, err := dsInfo.Type(ctx)
+			if err != nil {
+				return nil, logger.LogNewErrorCodef(log, codes.Internal,
+					"failed to get the type of datastore %q when checking if the number of snapshots on the "+
+						"source volume reaches the configured maximum. err: %v", datastoreUrl, err)
+			}
+
+			var isGranularMaxEnabled bool
+			if dsType == types.HostFileSystemVolumeFileSystemTypeVsan {
```

Comment:
I am not too familiar with this, but I believe there is vsan direct type as well https://opengrok.eng.vmware.com/source/xref/main.perforce.1666/bora/vim/vmodl/vim/host/FileSystemVolumeInfo.java#178 , for our our case do we check for both types?

Reply:
Currently, we only plan for VSAN and VVOL. From my understanding, we don't need to. VSAN Direct are basically VMFS datastore with lifecycle and healthy monitoring support from HCI-d. Anyway, we can add a config param later for VSAN Direct if requested by users.